### PR TITLE
[WIP][Code Quality] Lambda/API deployment

### PR DESCRIPTION
**Pull request by sidddutta**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: jenkins-build-pack-lambda/Jenkinsfile [hunk 1]
```diff
@@ -711,6 +731,28 @@ def getLambdaARN (String stackName) {
...
 	return ARN;
 }
+
+/** Run validation based on runtime
+*/
+def runValidation(String runtime) {
+	if(runtime.indexOf("nodejs") > -1) {
+		echo "running validations for $runtime"
+		sh "jshint *.js"
+	}else if(runtime.indexOf("java") > -1) {
```
Comment (svidhani) [inline jenkins-build-pack-lambda/Jenkinsfile]: indentation here and other places in the code

File: jazz-build-module/sonar-module.groovy [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+@Field def g_sonar_sources = "."
+@Field def g_sonar_java_binaries = "target/"
+@Field def g_sonar_host_url = ""
+@Field def g_sonar_login = ""
+@Field def g_sonar_password = ""
+@Field def g_key_prefix = "jazz"
+@Field def g_sonar_project_properties = [:]
+@Field def g_isVSScanEnabled = false
+@Field def g_dependencyCheckOutputFileName = "dependency-check-report.xml"
+@Field def g_dependencyCheckNISTFilesLocation = env.DEPENDENCY_CHECK_NIST_FILES_LOCATION
```
[the remark this answers is not in the document]
Comment (sidddutta) [reply jazz-build-module/sonar-module.groovy]: This is read from Config below.

File: jazz-build-module/sonar-module.groovy [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+@Field def g_sonar_java_binaries = "target/"
+@Field def g_sonar_host_url = ""
+@Field def g_sonar_login = ""
+@Field def g_sonar_password = ""
+@Field def g_key_prefix = "jazz"
+@Field def g_sonar_project_properties = [:]
+@Field def g_isVSScanEnabled = false
+@Field def g_dependencyCheckOutputFileName = "dependency-check-report.xml"
+@Field def g_dependencyCheckNISTFilesLocation = env.DEPENDENCY_CHECK_NIST_FILES_LOCATION 
+@Field def g_NISTDataMirrorUtility = env.DEPENDENCY_CHECK_NIST_MIRROR_UTILITY   //https://github.com/stevespringett/nist-data-mirror/releases/download/1.1.0/nist-data-mirror.jar
```
[the remark this answers is not in the document]
Comment (sidddutta) [reply jazz-build-module/sonar-module.groovy]: This is also taken from config now

File: jazz-build-module/sonar-module.groovy [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+import groovy.json.JsonOutput
+import groovy.transform.Field
+echo "Sonar code analyzer module loaded successfully"
+
+/**
+ * The Sonar module to run static code analysis - for JAVA, NodeJs & Python
+*/
+
+@Field def configLoader
+@Field def g_sonar_enabled = false // Note: set environment variable ENABLE_SONAR=true to enable sonar code analysis
```
[the remark this answers is not in the document]
Comment (sidddutta) [reply jazz-build-module/sonar-module.groovy]: This is also taken from config now

File: jazz-build-module/sonar-module.groovy [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+			g_sonar_profile =  sonarProfile 
+		}else{
+			g_sonar_profile = ""
+		}
+
+		setProjectKey(projectKey)
+		setProjectName(projectKey)
+		setRuntimeType(runtime)
+		
+		if(env.ENABLE_SONAR && env.ENABLE_SONAR == "true") {
```
[the remark this answers is not in the document]
Comment (sidddutta) [reply jazz-build-module/sonar-module.groovy]: This is also taken from config now

File: jenkins-build-pack-api/Jenkinsfile [hunk 6]
```diff
@@ -158,6 +161,23 @@ node() {
 			events.sendCompletedEvent('VALIDATE_PRE_BUILD_CONF', 'pre-build validation completed', context_map, environment_logical_id)
 		}
 
+		stage ('Code Quality Check'){
+			events.sendStartedEvent('CODE_QUALITY_CHECK', 'code quality check starts', context_map, environment_logical_id)
+			try {
+				clearVirtualEnv()
+				runValidation(runtime)
+
+				sonarModule.configureForProject(service, domain, branch, runtime, SONAR_JAZZ_PROFILE, isVSScanEnabled)
```
[the remark this answers is not in the document]
Comment (sidddutta) [reply jenkins-build-pack-api/Jenkinsfile]: A new initialize method created and fields are initialized from there.

File: jenkins-build-pack-api/Jenkinsfile [hunk 7]
```diff
@@ -661,6 +676,34 @@ def send_status_email (config, build_status, email_content) {
...
+/** Run validation based on runtime
+*/
+def runValidation(String runtime) {
+	if(runtime.indexOf("nodejs") > -1) {
+		echo "running validations for $runtime"
+		sh "jshint *.js"
+	}else if(runtime.indexOf("java") > -1) {
+		echo "running validations for $runtime"
+		dir('_java') {
+			sh 'wget -q $CHECKSTYLE_LIB'
```
Comment (suryajak) [inline jenkins-build-pack-api/Jenkinsfile]: this value will be from configLoader

File: jenkins-build-pack-lambda/Jenkinsfile [hunk 8]
```diff
@@ -711,6 +727,24 @@ def getLambdaARN (String stackName) {
...
+		echo "running validations for $runtime"
+		dir('_java') {
+			sh 'wget -q $CHECKSTYLE_LIB'
```
Comment (suryajak) [inline jenkins-build-pack-lambda/Jenkinsfile]: this value will be from configLoader

File: jenkins-build-pack-api/Jenkinsfile (hunk 7 above)
Comment (SiniWilson) [reply jenkins-build-pack-api/Jenkinsfile]: done

File: jenkins-build-pack-lambda/Jenkinsfile (hunk 8 above)
Comment (SiniWilson) [reply jenkins-build-pack-lambda/Jenkinsfile]: done

File: jenkins-build-pack-lambda/Jenkinsfile (hunk 1 above)
Comment (SiniWilson) [reply jenkins-build-pack-lambda/Jenkinsfile]: done

**Pull request merged**